aiosmtpd/tests/test_smtp.py

```diff
@@ -694,6 +714,63 @@ def test_unknown_command(self):
                 response,
                 b'Error: command "FOOBAR" not recognized')
 
+    def test_xclient_feature_not_enabled(self):
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('XCLIENT ADDR=10.1.1.1')
+            self.assertEqual(code, 550)
+            self.assertEqual(response, b'feature not enabled')
+
+
+class TestXCLIENT(unittest.TestCase):
+    def setUp(self):
+        self.handler = XCLIENTHandler(True)
+        controller = DecodingController(self.handler)
+        controller.start()
+        self.addCleanup(controller.stop)
+        self.address = (controller.hostname, controller.port)
+
+    def test_help_with_xclient_enabled(self):
+        """ Should return XCLIENT command if handler has hook """
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('HELP')
+            self.assertEqual(code, 250)
+            self.assertEqual(response,
+                             b'Supported commands: DATA EHLO HELO HELP MAIL '
+                             b'NOOP QUIT RCPT RSET VRFY XCLIENT')
+
+    def test_help_xclient(self):
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('HELP', 'XCLIENT')
+            self.assertEqual(code, 250)
+            self.assertEqual(
+                response, b'Syntax: XCLIENT [key1=value1] ... [keyN=valueN]')
+
+    def test_unknown_xclient_attribute_name(self):
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('XCLIENT FOOBAR=SPAM')
+            self.assertEqual(code, 501)
+            self.assertEqual(
+                response, b'FOOBAR is not a valid XCLIENT attribute')
+
+    def test_wrong_xclient_attribute_syntax(self):
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('XCLIENT FOOBARSPAM')
+            self.assertEqual(code, 501)
+            self.assertEqual(response, b'bad command parameter syntax')
+
+    def test_return_greeting_on_success(self):
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('XCLIENT ADDR=10.1.1.1')
+            self.assertEqual(code, 220)
+            self.assertEqual(response, b'success')
+
+    def test_return_handler_error(self):
+        self.handler.result = False
+        with SMTP(*self.address) as client:
+            code, response = client.docmd('XCLIENT ADDR=10.1.1.1')
+            self.assertEqual(code, 501)
+            self.assertEqual(response, b'bad command parameter syntax')
+
```

Comment:
What I don't see being tested is the result of a successful `XCLIENT` command.

The documentation for `XCLIENT` stated that this command

>  override the SMTP server's idea of the SMTP client hostname, network address, and other client information

which means -- and CMIIW here -- that the `envelope` or `session` information changes.

If so, then it should be tested / verified by having a successful `XCLIENT` followed by, say, a `MAIL FROM` and verifying that session/envelope changes accordingly.

Again, CMIIW; I'm not really familiar with `XCLIENT` command.